File: Leetcode/222-Count-Complete-Tree-Nodes/CountCompleteTreeNodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int countNodes(TreeNode* root) {
        int h = height(root);
        return h < 0? 0:
            height(root->right) == h-1?
                (1<<h) + countNodes(root->right) :
                (1<<h-1) + countNodes(root->left);
    }
private:
    int height(TreeNode* node) {
        return node == NULL? -1 : 1+height(node->left);
    }
};

File: Leetcode/222-Count-Complete-Tree-Nodes/CountCompleteTreeNodes.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def countNodes(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        h = self.height(root)
        if h < 0:
            return 0
        else:
            if self.height(root.right) == h-1:
                return (1<<h) + self.countNodes(root.right)
            else:
                return (1<<h-1) + self.countNodes(root.left)

    def height(self, node) :
        if node == None:
            return -1
        else:
            return 1 + self.height(node.left)
        
